### ISSUE TITLE
SqliteException(1): cannot rollback - no transaction is active

### ISSUE THREAD

**Issue by dutsky** [simolus3/drift#1882]
Hello! I'm using Drift 1.6.0 on Android with DriftIsolate. I'm trying to catch 'disk is full' exception(SqliteException(13)) when inserting chunks of data with batch.

Unfortunately, I can't catch it with try .. catch. My code looks like this:

```dart
@override
Future<void> insertThings(Iterable<Thing> things) async {
  try {
    await _db.batch((batch) => batch.insertAll(_db.things, things));
  } on SqliteException catch (e) {
     debugPrint('Exception happens: $e'));
  }
}
```

When disk becomes full, this unhandled SqliteException pops up:
`SqliteException(1): cannot rollback - no transaction is active`

```
E/flutter ( 5026): [ERROR:flutter/lib/ui/ui_dart_state.cc(198)] Unhandled Exception: SqliteException(1): cannot rollback - no transaction is active
E/flutter ( 5026):   Causing statement: ROLLBACK TRANSACTION
E/flutter ( 5026): #0      DriftCommunication.request
package:drift/…/remote/communication.dart:109
E/flutter ( 5026): #1      _RemoteTransactionExecutor._sendAction
package:drift/…/remote/client_impl.dart:181
E/flutter ( 5026): #2      _RemoteTransactionExecutor.rollback
package:drift/…/remote/client_impl.dart:189
E/flutter ( 5026): #3      Batch._commit
package:drift/…/api/batch.dart:188
E/flutter ( 5026): <asynchronous suspension>
```

After some investigation it looks like it's possible to handle exception by changing my code like this:
```dart
@override
Future<void> insertThings(Iterable<Thing> things) async {
  await _db.batch((batch) => batch.insertAll(_db.things, things))
  .catchError((Object e) => debugPrint('Exception happens: $e')));
}
```

but unfortunately I still can't catch 'disk is full' exception.

**Comment by simolus3** [comment]
Do you think there's some artificial way for me to reproduce this? I think you can't catch the `SqliteException` directly since it's happening on a remote isolate, but ideally drift should give you a wrapped `DriftRemoteException` for it. So the general `catchError` still doesn't catch the exception?

**Comment by dutsky** [comment]
I'm sorry, with `catchError` I can catch `SqliteException(1): cannot rollback - no transaction is active`, but can't catch `SqliteException(13) disk is full`.

I'm not sure about artificial way how to recreate it. I'm just used emulator with full /data directory and did some batch inserts, but I will think about it.

Thank you

**Comment by simolus3** [comment]
I could reproduce this locally by using deferred foreign key constraints (so that I can do an insert in a transaction that won't throw an exception right away but causes the commit to fail).

There were a bunch of problems here. In particular, [this method](https://github.com/simolus3/drift/blob/87848e58dd4071e47b880282a1e78446725bdc04/drift_sqflite/lib/drift_sqflite.dart#L94-L104) has pretty much been like this for a very long time and swallows all errors thrown by sqflite around `COMMIT` or `ROLLBACK` statements. I've also refactored how drift's transaction API maps onto sqflite's API to make sure that errors during all stages of the transaction are reported. If you want to, you can add this to your pubspec and check if this also helps for full disks. If not, I'll re-open and take another look :)

```yaml
dependency_overrides:
  drift:
    git:
      url: https://github.com/simolus3/drift.git
      ref: develop
      path: drift
  drift_sqflite:
    git:
      url: https://github.com/simolus3/drift.git
      ref: develop
      path: drift_sqflite
```

**Comment by TaifRaoof** [comment]
Hi, 
i still have the same problem. i am using the last version  2.2.0.

**Comment by simolus3** [comment]
@TaifRaoof Do you have more details on how I might reproduce this? Something like a log of statements or which calls you've made before this error?

**Comment by TaifRaoof** [comment]
this is my logs below

<img width="1300" alt="image" src="https://user-images.githubusercontent.com/11135440/201524832-4ab43831-fadd-4392-ac2e-19bfc6a0d576.png">

**Comment by TaifRaoof** [comment]
i solve this issue 
before :
Future<void> insertA(List<A> data) async => await batch((batch)=> batch.insertAll();
Future<void> insertB(List<B> data) async => await batch((batch)=> batch.insertAll();
 
call these function by
insertA();
insertB();


**After :** 
await insertA();
await insertB();


**i added await before each call function.**

**Comment by simolus3** [comment]
@TaifRaoof That is actually a bug in drift! You should always await statements of course, but drift shouldn't do anything stupid if you forget to do it either. In this case it mixed up two transactions, which is pretty bad. I have fixed this issue in f6830bcd68afecfd51aafc1462da796f14be1677. So thanks for reporting this!
